expo/expo
# [docs] Upgrade to next 9.5 and clean up legacy

# Why

This PR replaces #8651 and upgrades the docs to Next 9.5+ and cleans up custom legacy implementations. Also, see the optional changes that we could also do. I think they all help but aren't required for the Next 9 upgrade + legacy cleanup.

# How

## Changed

- Upgraded to Next 9.5+ using the [upgrade guide](https://nextjs.org/docs/upgrading).
- Moved `static` and `robots.txt` to `public/` to use [new public folder support](https://nextjs.org/blog/next-9-1#public-directory-support).

## Removed

- Legacy `@zeit/next-css` plugin, to [use built-in support](https://github.com/vercel/next.js/blob/master/errors/built-in-css-disabled.md).
- Legacy `server.js` and path map manipulation to fully rely on [`trailingSlash` config](https://nextjs.org/docs/api-reference/next.config.js/trailing-slash).
- Custom `babel-plugin-module-resolver` plugin, to [use built-in support](https://nextjs.org/docs/advanced-features/module-path-aliases).

## Fixed

- Replaced `/api/` page redirect with `pages/_error` redirect, to [avoid Next handling it as API endpoint](https://nextjs.org/docs/api-routes/introduction).
- Added workaround for the exported 404-fallback issue when [using `trailingSlash` config](https://github.com/vercel/next.js/issues/16528). → _note: once this is fixed, we can remove the fallback workaround from `package.json`_

## Optional

#### Symlink instead of copy `pages/versions/latest` folder → **see [this branch](https://github.com/expo/expo/compare/@bycedric/docs/next-9...@bycedric/docs/next-9-latest-link?expand=1#commits_bucket)**

This would speed up the build by avoiding ±13% (±150 files) of the total pages built. Still have to check with @nicknovitski what the reason for this workaround was. Looks like symlinks are properly resolved now in Next 9.5+.

#### Fixed anonymous component and `pages/_error` vs `pages/404` warnings → **see [this branch](https://github.com/expo/expo/compare/@bycedric/docs/next-9...@bycedric/docs/next-9-fix-warnings?expand=1#commits_bucket)**

Also, see [this blogpost](https://nextjs.org/blog/next-9-3) or run `$ yarn dev` and walk through the docs. These warnings are just warnings and should be visible in the CLI.

#### Replace global support for preval and scope it to `./constants` only → **see [this branch](https://github.com/expo/expo/compare/@bycedric/docs/next-9...@bycedric/docs/next-9-scoped-preval?expand=1#commits_bucket)**

This scopes preval to only the files that should require this. It's a "mental" limitation to avoid the [cache invalidation issue we currently have with preval](https://github.com/kentcdodds/babel-plugin-preval/issues/19). With this, we can cache the `.next/cache` folder, as [described in the docs](https://github.com/vercel/next.js/blob/canary/errors/no-cache.md), and remove `.next/preval` when running `$ yarn dev` to clean up the preval-ed files.

#### ~~Separate Next components from building tools or scripts, using the `./src` folder~~

~~This would clean up the docs folder and keep stuff like `./common/md-loader` out of Next's content scope, making it a bit more readable.~~ Requires more refactoring than I thought. If we want this, we need to update all paths referring to the GH docs files and references within expotools to any docs folder.

# Other stuff

## Preval vs Webpack plugin

There alternatives to preval, to avoid the cache invalidation issue currently present in preval. Unfortunately, it looks like they intend to deprecate these in favor of preval (e.g. [val-loader](https://github.com/webpack-contrib/val-loader/issues/38#issuecomment-546946385)). That means we probably have to stick with preval.

We could also refactor everything to use `getStaticProps`, but that might take some time. Currently, we use the data from preval as global shared constants. Data from `getStaticProps` is only available in pages.

## Performance

Unfortunately, there is still a small performance degradation when upgrading to Next 9.5+. Luckily it is not as much as mentioned in #8651 (±3x).

action | Next v8 | Next v9 + MDX v0 | Next v9 + MDX v1
--- | --- | --- | ---
build _without_ cache | `317.43s` | `466.33s` | `370.73s`
build _with_ cache | `97.51s` | `87.71s` | `101.99s`

> These are the averages of running 3x using a shell script, with the `/usr/bin/time` tool.

# Test Plan

- Run docs locally through `$ yarn dev` and try stuff out 😄 
- Run `$ yarn export` and `$ yarn export-server` to see the exported version
